### PR TITLE
Selective UIA event handling

### LINKED ISSUE #11077: Restrict most event processing to objects of interest to the user
### problem
Some web apps, such as Twitter and Slack, fire a lot of events, most of which aren't things we need to report to the user. For example, when you focus the message list in Slack, it changes the name for every item in the list and adds the Message actions grouping to every item in the list. This results in noticeable performance problems. On my system with Firefox Nightly, focusing the Slack message list with 32 items in the list takes > 700 ms to respond. This is quite a bit faster in Chrome, so clearly there's some work to do in Firefox (and I'm looking into that as well). However, processing less of these events makes things significantly faster (~500 ms or less with my prototype).

We've had similar problems with UIA. For example, 1Password fires many property change events when you switch items in the list (#10508).

### High Level Solution
NVDA already drops events for background windows, etc. However, there can be a lot of events even in the foreground window that aren't important to the user. For example, a nameChange on an object that isn't focused and isn't in the focus ancestry isn't reported in any way. Previously, these events were processed, NVDAObjects were instantiated for them and the events were queued, but they simply did nothing when they were finally handled by the NVDAObject.

To address this, I propose that NVDA only process most IAccessible events for the focus, focus ancestry, navigator object and desktop object. For UIA, we could do the same, but it probably makes sense to only listen for events on specific objects; see below.

### Issues
I've prototyped two versions of this for IAccessible.

1. Do all of this in IAccessibleHandler. Problems:
    - This doesn't integrate well with eventHandler.requestEvent. If a plugin requests an event, it will get dropped before shouldAcceptEvent is called. And we can't do this stuff after shouldAcceptEvent because it will return True for cases we want to block.
    - We'd need a completely separate implementation for UIA; we can't use any common code.
2. Integrate it into shouldAcceptEvent. Pass raw API event params to shouldAcceptEvent and let it ask the relevant NVDAObjects whether the event is for them.
    - The problem is that for UIA, it makes more sense to just only listen for events on relevant objects, rather than listening for all events and dropping the ones we don't care about. In contrast, we can't specify specific objects for IAccessible.

It's really hard to come up with a framework that works well for both. On the other hand, as can be seen from the complexity of shouldAcceptEvent, duplicating all of the logic between IAccessible and UIA wouldn't be sustainable.

I think addressing this for both APIs is probably going to involve a hybrid approach. We'll need some common logic in eventHandler, but this will need to be split up so we can call different pieces in different places. For example, we need to be able to separately ask "did a plugin request this?" and "is this event for the foreground window?", rather than this all being handled in one call. We'll also need to be able to ask "what objects are of interest to the user?", since IAccessible and UIA will handle that differently (filtering vs selective registration).

@michaelDCurran, @feerrenrut, thoughts appreciated. I can push both of my prototype patches somewhere if you're interested (draft pull requests perhaps?), but I didn't want to spam with in-progress work without asking/discussing first.

### LINKED ISSUE #11209: UIA: selective event registration
Related to #11077, #11109, #11002, #8742 

### Is your feature request related to a problem? Please describe.
NVDA is listening globally to many UIA elements. This causes a major performance hit when many events are received (#11109)

As part of #11077, I investigated a suggestion by @jcsteh to only queue events for objects of interest. However, this doesn't cause the major performance improvement I was hoping for.

### Describe the solution you'd like
I'm going to prototype the following:

1. An easier method for event registration by means of EventHandlerGroups. This feature was introduced in IUIAutomation6. I already wrote a FakeEventHandlerGroup class to mimic this behavior on older versions of UIA. In short, this should allow us to request for automation/ notification and property events in one call instead of three on newer versions of UIA.
2. When changing focus, register for events fired at the focus object and its ancestors.
3. distinguish between events that are of global interest (notifications, focus changes, value changes) and local interest (name, description, item status, etc.). We can have a global and a local event handler group. The global group will be registered once, the local group will be added/removed on focus.

Nothing is set in stone yet. I just want to tackle these issues as they have major impact on my daily work.

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Closes #11209 
fixes #11077
Improves #11002
Follow up of #8742 

### Summary of the issue:
NVDA is listening globally to many UIA elements. This causes a major performance hit when many events are received (#11109)

### Description of how this pull request fixes the issue:
1. This pr introduces an easier method for event registration by means of EventHandlerGroups. This feature was introduced in IUIAutomation6. I wrote a FakeEventHandlerGroup class to mimic this behavior on older versions of UIA. In short, this allows us to request for automation/ notification and property events in one call instead of three on newer versions of UIA.
2. For some events, we still keep listening globally. However, for many of them, I switched to local event registration, with that I mean only to objects of interest. Currently, these are the focus object and its ancestors.
3. When changing focus, we register for events fired at the focus object and its ancestors. This includes all focus changes, including these for non native UIA elements. This includes focus changes for IAccessible objects as well.

### Testing performed:
* @codeofdusk confirmed that #11002 has improved by this
* I myself can confirm that the Nuget Package manager as well as other aspects of Visual Studio are much snappier.

### Known issues with pull request:
* @josephsl noted that announcements for modern input are somewhat broken, i.e. the announcement that the emoji panel is closed. I haven't yet looked into this. I'd like to have a basic review about the prototype and its performance improvements first.

### Change log entry:
* Bug fixes
    + Major performance improvements in Microsoft Visual Studio and other UIAutomation based applications. (#11077, #11209)

### CODE REVIEW

**Comment by leonardder** [comment]
[Here is a try build](https://ci.appveyor.com/api/buildjobs/fkleiquur2rw78c2/artifacts/output%2Fnvda_snapshot_try-selectiveUIAEventHandling-20244%2Ccd4a3254.exe)

**Comment by lukaszgo1** [comment]
@leonardder I can confirm that the latest try build fixes my issues under Windows 7.

**Comment by josephsl** [comment]
Hi,

Regarding emoji panel and other modern input features, the fundamental issue to consider is focus change. A few months ago I tried to resolve overlay focus issues by faking a focus change event when emoji panel opens, but this resulted in NVDA's notion of "focus" being trapped once the panel was closed i.e. previously focused window isn't announced once the panel closes. In theory, this issue would also impact touch keyboard support on Windows 8 and later if it also suffers from focus trap issue, along with a possible impact on UIA notification event handling for classic Edge. I'm looking into resolving emoji panel issue from the app module side and to test this PR at the same time.

Thanks.

**Review by josephsl — COMMENTED** [review]
It turns out focus change handler will add or remove events every time focus changes. This is more so when moving focus to a different element inside the same foreground window/application. This might be something to optimize later.

Apart from modern input support not working well, this looks promising. Thanks.

**Comment by josephsl** [comment]
Hi,

Found a way to restore emoji panel support - local event group must be added for input panel header. This means this PR must block #10378 - I'll package that PR and the follow-up one together.

Thanks.

**Comment by codeofdusk** [comment]
Found a bug:

1. in 21H1 (open) console with UIA, press alt+space to open the application menu.
2. Press e to open the edit menu.
3. Press p to paste.

Expected: auto-read continues functioning after this, with no need to switch windows.
Actual: a focus change is necessary to bring back autoread.

**Comment by leonardder** [comment]
@josephsl wrote:

> It turns out focus change handler will add or remove events every time focus changes. This is more so when moving focus to a different element inside the same foreground window/application. This might be something to optimize later.

This is currently intentional. #8742 implemented per foreground event registration, but as @jcsteh noted in https://github.com/nvaccess/nvda/pull/8742#issuecomment-616868542 , it wouldn't fix property event bombarding within the same foreground window. 

@codeofdusk wrote:
> Actual: a focus change is necessary to bring back autoread.

This is really weird. It looks like the local event registration isn't triggered when focus is back on the terminal window. I will investigate this.

**Comment by josephsl** [comment]
Hi, I see. By the way, for modern input, one must first remove local events before adding local event handler group, otherwise event groups will get large, leading to NVDA announcing things multiple times. This must be done for item selected and window open event – the former for emoji panel, the latter for dictation and hardware keyboard input, respectively. Thanks.

**Comment by leonardder** [comment]
Ugh, so UIA allows you to register one handler to elements multiple 
times? That's horrific!

I think we need something like eventHandler.requestEvents for UIA. If 
I'm correct, for modern input, we only need the name change anyway.

**Comment by josephsl** [comment]
Hi, yes for emoji panel, item selected event may be needed for hardware keyboard input. Eventually we need to do something about clipboard history context menu, but that’s beyond this PR. Thanks.

**Comment by leonardder** [comment]
item selected event is listened to globally, as it is also required for 
search results.

**Comment by leonardder** [comment]
@codeofdusk It looks like the issue with the UIA implementation of cmd is exactly the same issue as that requires us to implement focusRedirect. I will investigate a fix.

**Comment by codeofdusk** [comment]
More generally, I'd think you'd want to find a method of dealing with focus changes through redirects...

**Comment by leonardder** [comment]
I will investigate whether it is possible to hide stuff behind a feature flag, so the initial version of this pull request can hopefully be merged sooner, with may be some issues remaining.

**Comment by feerrenrut** [comment]
@codeofdusk No, given this is still a draft and introduces lots of changes to UIA I think it would be safer to target 2020.3

**Comment by leonardder** [comment]
Agreed. Also, not that i don't trust @feerrenrut  here, but I really want a look from @michaelDCurran as well.

**Comment by feerrenrut** [comment]
Yes, given the nature of the code, I think a double review is wise.

**Comment by leonardder** [comment]
> @codeofdusk No, given this is still a draft and introduces lots of changes to UIA I think it would be safer to target 2020.3

While I still agree, I'd like to note that I've been using this for three weeks consecutively without major issues, other than the Windows command line @codeofdusk pointed us at. Therefore, I think it would really help if we could explicitly target 2020.3 (i.e. not having it delayed to a later version than that). Therefore the positive impact is too big to delay it ffurther.

**Comment by josephsl** [comment]
Hi, I’ll continue to investigate modern input issues in the meantime. Thanks.

File: source/_UIAHandler.py
Comment (michaelDCurran) [inline source/_UIAHandler.py:487]: Why does this have to be queued?  the focus event handler is already being executed in an MTA thread (granted it is probably not our specific one, but it is certainly running within the MTA apartment). Did you see problems not doing this?

File: source/_UIAHandler.py
Comment (michaelDCurran) [inline source/_UIAHandler.py:306]: I certainly understand the advantage of using a local event handler group as we will be registering and unregistering a lot. But is there an actual advantage to using a global event handler group here? Sure it will possibly speed up UIA initialization a little I guess... any other reason?

**Comment by michaelDCurran** [comment]
I'm guessing the troubles with the Windows console after exiting a menu is caused by the fact that the terminal control itself never fires a UIA focus event when the menu exits. The only reason NVDA thinks there is a focus change is because it detects a menu end event in IAccessible and fakes a focus event, I think.  It might be better to register and deregister local event handler groups in UIA NVDAObject's event_gainFocus?  I understand that in that case you may miss registering for some UIA ancestors if the focus itself ended up being handled with MSAA... but I'm not sure how common that particular scenario is. Another option is to listen for UIA_menuClosedEvent globally in UIA, and correct the focus by calling IUIAutomationFocusChangedEventHandler_HandleFocusChangedEvent directly giving it the element returned from UIAutomation's GetFocusedElement.

File: source/_UIAHandler.py
Comment (michaelDCurran) [inline source/_UIAHandler.py:184]: I think it is possibly better to keep this global rather than local. Selection tracking is rather like focus tracking in that there usually are not too many events, and they can come from anywhere. In fact the most important cases, such as with search suggestions, and I'm guessing things like emoji pannel:  the selected items have nothing at all to do with the focus.

File: source/_UIAHandler.py
Comment (leonardder) [reply source/_UIAHandler.py:487]: Honestly, I copied this behavior from #8742, but yeah, this implementation is different so that should work. I'll give it a spin.

File: source/_UIAHandler.py
Comment (leonardder) [reply source/_UIAHandler.py:306]: I had two reasons for it:

1. Consistency. If we do event registration one way for local events, it made sense to me to do it exactly the same way for the global events
2. According to [Microsoft Docs](https://docs.microsoft.com/en-us/windows/win32/api/uiautomationclient/nf-uiautomationclient-iuiautomation6-addeventhandlergroup), this is the preferred way of registration:

    > Microsoft UI Automation clients should use the handler group methods to register event listeners instead of individual event registration methods defined in the various IUIAutomation interface namespaces.

**Comment by leonardder** [comment]
> Another option is to listen for UIA_menuClosedEvent globally in UIA, and correct the focus by calling IUIAutomationFocusChangedEventHandler_HandleFocusChangedEvent directly giving it the element returned from UIAutomation's GetFocusedElement.

I tested that approach, and it didn't work, so I'll go investigate the first one. That also requires us to keep the function queue to ensure we're registering in the right thread apartment.

**Comment by leonardder** [comment]
> It might be better to register and deregister local event handler groups in UIA NVDAObject's event_gainFocus? 

This makes perfect sense to me. I updated the pr accordingly, of course deregistration happens in loseFocus now. The elegance of this is that we're no longer registering for local events for non UIA objects at all. I assume this could be beneficial if one forcefully disables UIA for a certain control. The current approach can also be expanded to navigator objects if done properly.

> I understand that in that case you may miss registering for some UIA ancestors if the focus itself ended up being handled with MSAA... but I'm not sure how common that particular scenario is.

I know that Wordpad is one case where this applies, however I guess it is very uncommon that we'd need events from one of the ancestors in these cases. @josephsl could you think of any?

**Comment by josephsl** [comment]
Hi, I’m not sure at this point – I’ll need to test latest commit from this PR with emoji panel, Windows Terminal and other scenarios at some point. Thanks.

**Comment by michaelDCurran** [comment]
it will be very important for a try build of this to be thoroughly tested in Windows 7 again, as the way that UIA events are registered has changed again in this pr. When I was originally developing support for UIA back in 2009 for Windows 7, I was getting freezes all the time with this approach. I'm fairly confident that the freezes won't exist in Windows 10, and they are probably addressed in Windows 7 by platform updates, but still important to ensure things are okay for such a big change like this. Though I do hope we can merge this pr once we know it is stable.

**Comment by zstanecic** [comment]
Why to test this in windows 7, as it lost support?

**Comment by josephsl** [comment]
Hi, because some enterprises would have subscribed to ESU (extended security updates), guaranteeing support until January 2023 (although it means paying Microsoft to keep computers secure). Thanks,

**Comment by leonardder** [comment]
> the way that UIA events are registered has changed again in this pr.

Yes and no. The way has changed, but in Windows 7, we're using our own clone of EventHandlerGroups. Under the hoot, it should still translate to the same behavior.
Of course, major changes apply when the feature flag in NVDA's advanced settings is enabled, and I second the request to test this.

@lukaszgo1 would you be able to test this during daily usage? I will provide a try build that has all fixes incorporated.

**Comment by josephsl** [comment]
Hi, I might be able to assist also – let us know which scenarios you want us to test and what’s expected so we can report our findings. Thanks.

**Comment by leonardder** [comment]
@josephsl I think it would help to test any differences in behavior on Windows 7 with and without this feature enabled, preferably in primarily UIA driven applications like Visual Studio Code, Outlook/Word with UIA enabled.

**Comment by josephsl** [comment]
Hi, the thing is I don’t use any of these on my Windows 7 VM, and I’ll be testing with networking off for security reasons. Thanks.

**Comment by lukaszgo1** [comment]
@leonardder wrote:
> @lukaszgo1 would you be able to test this during daily usage? I will provide a try build that has all fixes incorporated.


Yes. Would testing with the try build as a portable be sufficient, or is installation necessary in your opinion?

also:

> I think it would help to test any differences in behavior on Windows 7 with and without this feature enabled, preferably in primarily UIA driven applications like Visual Studio Code, Outlook/Word with UIA enabled.

That is going to  be problematic. I don't  have versions of Office higher than 2010 on any of my Windows 7 machines so I'm unable to test with UIA enabled. I'm also not a Visual Studio user (I believe you mend full VS rather than VS Code which  AFAIK is not UIA based). 
As an aside I'm really curious how other AT's i.e. JAWS and Window-eyes back in the day when WE was a think solved this - they never suffered from these UIA slowness regardless of the Windows version in use. Too bad we would probably never know what solution they've found.

**Comment by leonardder** [comment]
> I believe you mend full VS rather than VS Code

Correct, I"m sorry.

**Comment by codeofdusk** [comment]
If most of these changes are behind a feature flag, couldn't we just merge this and have win7 users test before we enable it by default?

**Comment by leonardder** [comment]
Most recent try build lives at https://ci.appveyor.com/api/buildjobs/yaqii668ensbceia/artifacts/output%2Fnvda_snapshot_try-selectiveUIAEventHandling-20406%2Cd18670c1.exe

**Comment by josephsl** [comment]
Hi, so far, I’m not noticing any differences on Windows 7 (note that I’m testing with try build set to portable copy and run as admin). I’m also testing on Windows 8.1 and Windows 10 (all feature updates throughout history). Thanks.

**Comment by josephsl** [comment]
Hi,

By the way, in order to properly test this, one must:

1. Change selective UIA events setting from Advanced settings panel.
2. Reload UIA handler thread as follows:

Open Python console, then:
import UIAHandler
UIAHandler.terminate(); UIAHandler.initialize()

The import statement is not needed if UIA handler was already imported.

Thanks.

**Comment by josephsl** [comment]
Hi,

My results so far:

* Windows 7 and 8.1: no difference observed, although this was with no UIA intensive apps such as Microsoft 365 installed.
* Windows 10 versions 1507 through 1709: no difference I can observe.
* Windows 10 Version 1803: no hardware keyboard suggestions are announced when selective UIA events is on.
* Windows 10 Version 1809 and later: no announcement of emoji panel categories when pressing tab from emoji panel.

Scenarios tested:

* Opening, searching, and closing Start menu/screen.
* Opening and closing Control Panel.
* Opening, navigating, and closing Windows 10 Settings app.
* File Explorer.
* Modern input.

For hardware keyboard suggestions and emoji panel, one workaround is manually registering various UIA events for modern input apps. This does allow hardware keyboard suggestions to be announced and emoji categories to be heard when pressing Tab from emoji panel in 1809 and later.

Thanks.

**Comment by josephsl** [comment]
Hi,

Another thing to consider: in Windows 10's Start menu, after selecting a search result or moving focus away from the screen itself, suggestions closed sound is not heard (this is the case with at least 20H2).

Thanks.

**Comment by michaelDCurran** [comment]
@josephsl  What specific UIA events are required for hardware keyboard and emoji selection support that was broken with this pr?

**Comment by josephsl** [comment]
Hi, “window open” and “item selected”. Thanks.

**Comment by leonardder** [comment]
Personally, I think we shouldn't make these issues block this pr from going through, as the new behavior isn't the default. Also, in 2004 it works fine for me, at least when pressing escape to close the start menu.

**Comment by josephsl** [comment]
Hi, that I can confirm that. I think for now it would be preferable to let people test this for a long time and come up with edge cases as time goes. Thanks.

**Comment by zstanecic** [comment]
Yep, I agree.

The more the testing cases from master,the more the possiblility for fixing issues.

**Comment by zstanecic** [comment]
How is situation with this?

**Comment by leonardder** [comment]
i've been using this for the last month or so on a Windows 10 machine in production, as Visual Studio is almost unbearable without it.
Personally, if @michaelDCurran agrees with the current code, I'd appreciate a merge early in the 2020.3 alpha cycle so it can prove itself.

**Comment by zstanecic** [comment]
I am really for merging this,

It really proves the performance boost

**Comment by josephsl** [comment]
Hi, or rather, gives us plenty of time to test edge cases and come up with proper workarounds and could serve as a stepping stone for a bigger things. I second early merge. Thanks.

**Comment by michaelDCurran** [comment]
I will look into reviewing / merging this on Monday.

**Review by michaelDCurran — APPROVED** [review]

**Comment by codeofdusk** [comment]
This feature has been disabled by default for several releases.

At this stage, what is blocking enabling by default?

**Pull request merged**
